### PR TITLE
Fix accelerator conflict with apply button.

### PR DESCRIPTION
**Pull request by feerrenrut**
### Link to issue number:
#8249 

### Summary of the issue:
The apply button shares an accelerator (in English, "alt+a") with other controls, when this accelerator is triggered for a button the button is activated. The only feedback that a user gets that this is the case is that the focus shifts to the categories list.

Since ok can be triggered with enter, and cancel with esc, it seems that apply should also have a keyboard shortcut.

### Description of how this pull request fixes the issue:
One considered idea was to remove the apply button, however we expect that some users use the apply button to test out different settings without having to reopen the settings dialog and find the settings that they were up to again. In the future we may move to a settings dialog that allows you to preview settings as soon as they are changed, only committing to the change when ok or cancel. This would allow us to remove the apply button and still support this use-case.

This commit stops the settings dialogs from using the "CreateButtonSizer" method, and instead we create the buttons manually. We add a binding for the "enter" key so that enter will still trigger the OK button. We also add a binding for "control + S" to trigger the apply button.

Also included is a change to resolve the annoying windows chime when pressing enter on the synthesiser name, or braille display name, the "...Change" button is now
triggered.

Because we are creating these buttons manually, we are able to translate the OK, Cancel, Apply for all our users.

### Testing performed:
* Enter / esc from categories list still accepts / discards changes 
* Enter on a button (such as synthesizer change button) still triggers the button
* Enter / esc on a checkbox item still accepts / discards changes
* Enter in an expanded dropdown list selects the current item
* Ctrl + s from a checkbox item applies settings and moves focus to the categories list
* Ctrl + s from the categories list applies settings
* Ctrl + s from the ok/cancel/apply buttons does apply settings
* Ctrl + tab shifts categories
* Enter on the synth label opens change synth dialog
* Enter on the braille display label opens the change braille display dialog

### Known issues with pull request:
There is still no explicit notification that settings have been applied other than focus shifting to the categories list.

### Change log entry:
**Changes**
`Settings can now be applied by using 'control + s' in the NVDA settings dialog. (#8249)`

### CODE REVIEW

File: source/gui/settingsDialogs.py [hunk 1]
```diff
@@ -724,7 +747,8 @@ def makeSettings(self, settingsSizer):
 		settingsSizerHelper = guiHelper.BoxSizerHelper(self, sizer=settingsSizer)
 		# Translators: A label for the synthesizer on the speech panel.
 		synthLabel = _("&Synthesizer")
-		synthGroup = guiHelper.BoxSizerHelper(self, sizer=wx.StaticBoxSizer(wx.StaticBox(self, label=synthLabel), wx.HORIZONTAL))
+		synthBox = wx.StaticBox(self, label=synthLabel)
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: May I ask why this is changed? It seems this change hasn't been made to the BrailleSettingsPanel.

Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: Not a functional change, more of an incidental cleanup. I will make the same change in the BrailleSettingsPanel to ensure they are consistent.

**Comment by feerrenrut** [comment]
A try build is available for people to test this: https://ci.appveyor.com/api/buildjobs/pygko7p8vm544oj5/artifacts/output%2Fnvda_snapshot_try-i8249-replaceApplyButtonAndShortcuts-15115%2C165ed421.exe

We would like to merge this change in soon, so that translations can can be completed prior to release. Feedback is welcomed, however we will likely only block this change if a major issue is found.

**Review by michaelDCurran — APPROVED** [review]

**Comment by zstanecic** [comment]
HI mick,

after the testing,

can it be merged on the translation strings freeze day, as this brings 
the awaited change?

croatian will have translated buttons in the settings dialog.

can it also be affected for windows 10 ocr dialog?



W dniu 09.05.2018 o 06:50, Michael Curran pisze:

**Comment by michaelDCurran** [comment]
This will be merged on Friday at which time I will do Translation freeze.
We are not going to change any extra dialogs at this late stage.

**Comment by zstanecic** [comment]
yes, but this will be inconsistent then.

i mean, just push the change which will make ok and cancel translatable 
in the other dlg



W dniu 09.05.2018 o 09:42, Michael Curran pisze:

**Comment by leonardder** [comment]
[**@zstanecic**](https://github.com/zstanecic) commented on [9 mei 2018 08:41 CEST](https://github.com/nvaccess/nvda/pull/8252#issuecomment-387636548 "2018-05-09T06:41:55Z - Replied by Github Reply Comments"):
> can it also be affected for windows 10 ocr dialog?

I don't understand your request. There is no Windows 10 OCR dialog any more, as that dialog is now part of the NVDA Settings dialog, to which this change applies.

**Comment by feerrenrut** [comment]
>yes, but this will be inconsistent then.
>
> i mean, just push the change which will make ok and cancel translatable
in the other dlg

@zstanecic  Yes, this will be inconsistent. The change to make ok and cancel translatable comes with a real risk of breaking the dialog, it requires a lot of testing. It is too close to the release date to be able to do this for all dialogs.

**Comment by leonardder** [comment]
[**@zstanecic**](https://github.com/zstanecic) commented on [9 mei 2018 10:20 CEST](https://github.com/nvaccess/nvda/pull/8252#issuecomment-387661037 "2018-05-09T08:20:22Z - Replied by Github Reply Comments"):
> we have punctuation pronounciation dialog

This change also applies to that dialog. Except for the modal dialog that is used to create a new dictionary entry or add a symbol, all settings dialogs should be covered.

**Comment by Brian1Gaff** [comment]
Hi folks, this may or may not be related to this change but I chose to 
download this version and I am then, of course back to the old browse for a 
folder to make a portable version interface. It seems however that at least 
for me, the folder list  has every folder  spoken twice. This did not used 
to be the case at all. This is of course Windows 7 on my main machine. Its 
not a problem but  just rather odd. This may have crept in much earlier but 
of course the new update function bypasses this folder choice and assumes 
you want to create it in the folder of the running version so one never sees 
the folder display.

However back on topic for the  changes. I hear not real difference here to 
the old way, the synth can be changed as before and ok verifies it and you 
can still change other items in the dialogue and hit apply and it works.



 Brian

**Comment by zstanecic** [comment]
it is ok then



W dniu 09.05.2018 o 10:38, Leonard de Ruijter pisze:

**Comment by Brian1Gaff** [comment]
I guess a note in the whats new or user guide about this might be a good 
idea for non English users then to try to head of lots of tickets on the 
condition?

Brian

**Comment by Brian1Gaff** [comment]
In that special version I'm also getting double reading in IE and in any 
html based on the IE engine, ie Outlook Express, live mail etc.

Has something happened to cause this, as master looks to be unaffected.

Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message -----

**Comment by zstanecic** [comment]
i agree,

but it should be merged tomorrow or today, so that we have synchronized 
as translators



W dniu 09.05.2018 o 10:48, Brian Gaff pisze:

**Comment by leonardder** [comment]
@feerrenrut @michaeldcurran: It is actually possible from the appModule to set control+s as a shortcut on the apply button objects by means of the windowControlId, which is wx.ID_APPLY. However, this implies that every custom button that uses wx.ID_APPLY will have the control+s event listener on it. Within NVDA itself, that would be true, but probably not for add-ons, unless they inherit their dialogs from SettingsDialog.

**Pull request merged**